### PR TITLE
feat: add lightweight AssemblyScript SDK for Wagi HTTP components

### PR DESCRIPTION
**Pull request by radu-matei**
ref #298 

This commit starts the work for a lightweight Spin SDK
for AssemblyScript.
Particularly, it starts the work on defining the HTTP
objects needed, and adds a work-in-progress for reading
request data from the environment (for Wagi components).
This is just a work-in-progress, as reading the request
body and writing the response body are currently not implented.

With this SDK, writing an HTTP component in AssemblyScript:

```typescript
export function _start(): void {
    handleRequest((request: Request): Response => {
        for (var i = 0; i < request.headers.size; i++) {
            Console.error("Key: " + request.headers.keys()[i]);
            Console.error("Value: " + request.headers.values()[i] + "\n");
        }
        return new ResponseBuilder(StatusCode.FORBIDDEN)
            .header("content-type", "text/plain")
            .header("foo", "bar")
            .body(String.UTF8.encode("Hello, Fermyon!\n"));
    });
}
```

Note that this still follows the existing convention for
calling the `_start` function in the module for a Wagi
component.

(We are currently experimenting with going through the C bindings
for writing components in other languages, currently for CGO.
If that experiment is successful, it would allow us, for the list
of languages supported with an SDK, to stop relying on the Wagi
compatibility layer and fully implement all types of components.
That work is still early. Depending on its outcome, we will decide
how to approach the AssemblyScript SDK.)

Signed-off-by: Radu Matei <radu.matei@fermyon.com>

### CODE REVIEW

File: sdk/as/http/cgi.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    let method = requestMethodFromEnv();
+    let headers = requestHeadersFromEnv();
+    // let body = requestBodyFromEnv();
+
+    return new Request(uri, method, headers);
+}
+
+export function sendResponse(response: Response): void {
+    printHeaders(response);
+    // wasi.Descriptor.Stdout.write(changetype<u8[]>(response.body));
```
Comment (radu-matei) [inline sdk/as/http/cgi.ts:15]: Reading the request body is currently not working.

File: sdk/as/http/cgi.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
+import * as wasi from 'as-wasi';
+import { Method, Request, Response } from "./http";
+
+export function requestFromCgi(): Request {
+    let uri = requestUriFromEnv();
+    let method = requestMethodFromEnv();
+    let headers = requestHeadersFromEnv();
+    // let body = requestBodyFromEnv();
```
Comment (radu-matei) [inline sdk/as/http/cgi.ts:8]: Reading the request body is currently not working.

File: sdk/as/http/http.ts [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+            return Method.TRACE;
+        } else if (m == "PATCH" || m == "patch") {
+            return Method.PATCH;
+        } else {
+            return Method.GET;
+        }
+    }
+
+    /** Convert an HTTP method into a string. */
+    export function from(m: Method): string {
```
Comment (radu-matei) [inline sdk/as/http/http.ts:143]: Terrible function name.

File: sdk/as/http/http.ts [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    DELETE,
+    CONNECT,
+    OPTIONS,
+    TRACE,
+    PATCH,
+}
+
+export namespace Method {
+    /** Parse a string into an HTTP method. */
+    export function parse(m: string): Method {
```
Comment (radu-matei) [inline sdk/as/http/http.ts:118]: Terrible function name.

**Comment by radu-matei** [comment]
TODO:

- [ ] decide whether this SDK will still follow the Wagi model for HTTP components
- [ ] fix request and response bodies
- [ ] add outbound HTTP
- [ ] add tests

**Comment by lann** [comment]
I know `as` is a typical abbreviation in that ecosystem, but maybe give it a longer name here?

**Comment by itowlson** [comment]
As a heads up, one thing to test for in this is handling of the special `_routes` entry point.  They believe they have fixed the issue with non-default entry points, but from their description (https://github.com/AssemblyScript/assemblyscript/pull/2157) I'm not sure it will work for WAGI.

(And this may be on us.  WAGI doesn't really treat the module as a reactor, but instead kinda treats it as two commands in a trenchcoat, which works okay with Rust but might be a bit ambiguous in terms of the WASI spec, I'm not sure.)

**Comment by radu-matei** [comment]
Oh, this brings up an excellent point that we (read I) should have documented and/or implemented — see #342

**Comment by radu-matei** [comment]
AssemblyScript recently removed WASI support from the core project, and we no longer support the old WASI experimental HTTP library, so defering this SDK for now.

Happy to revisit if people are interested.

ref #699

**Pull request closed without merging**